### ISSUE TITLE
npc appears to be ignoring a networkpolicy for ingress traffic

### ISSUE THREAD

**Issue by deitch** [weaveworks/weave#2997]
## What you expected to happen?

traffic to be allowed through

## What happened?

traffic blocked

## How to reproduce it?

1. Create an ingress controller listening on 80
2. Create a `nodePort` service listening on 30080 forwarding to the ingress controller pods on 80
3. Try to connect from outside to 30080 on host, get tcp traffic blocked by npc
4. Create a networkpolicy to enable 80 traffic
5. Try to connect, traffic _still_ blocked by npc

## Anything else we need to know?

Kube 1.6.2, CoreOS 1353.8.0

## Versions:
```
$ weave version
1.9.6
$ docker version
Client:
 Version:      1.12.6
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   d5236f0
 Built:        Tue May 30 23:15:08 2017
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.6
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   d5236f0
 Built:        Tue May 30 23:15:08 2017
 OS/Arch:      linux/amd64
$ uname -a
Linux ip-10-78-22-156.ec2.internal 4.9.24-coreos #1 SMP Tue May 30 23:12:01 UTC 2017 x86_64 Intel(R) Xeon(R) CPU E5-2686 v4 @ 2.30GHz GenuineIntel GNU/Linux
$ kubectl version
1.6.2
```

I am using a traefik ingress controller and service:

```yml
apiVersion: v1
kind: Service
metadata:
  name: traefik-ingress-controller
  namespace: kube-system
  labels:
    app: traefik-ingress-controller
spec:
  type: NodePort
  ports:
  - port: 80
    name: http
    targetPort: 80
    nodePort: 30080
  - port: 443
    name: https
    targetPort: 443
    nodePort: 30443
  selector:
    name: traefik-ingress-lb
---
kind: DaemonSet
apiVersion: extensions/v1beta1
metadata:
  name: traefik-ingress-controller
  namespace: kube-system
  labels:
    name: traefik-ingress-lb
    kubernetes.io/cluster-service: "true"
spec:
  selector:
    matchLabels:
      name: traefik-ingress-lb
  template:
    metadata:
      labels:
        name: traefik-ingress-lb
    spec:
      terminationGracePeriodSeconds: 60
      containers:
      - image: traefik
        name: traefik-ingress-lb
        resources:
          limits:
            cpu: 200m
            memory: 30Mi
          requests:
            cpu: 100m
            memory: 20Mi
        ports:
        - containerPort: 80
          name: http
        - containerPort: 8080
          name: https
        args:
        - --web
        - --kubernetes
        - --kubernetes.filename=/etc/traefik/config/kubernetes.tmpl
        volumeMounts:
          - name: traefik-tmpl
            mountPath: /etc/traefik/config
            readOnly: true
      volumes:
        - name: traefik-tmpl
          configMap:
            name: traefik-tmpl
            items:
              - key: kubernetes.tmpl
                path: kubernetes.tmpl

```

Looking at one worker 10.78.22.156 and trying to `telnet 10.78.22.156 30080` from 10.78.11.124:

```
WARN: 2017/05/31 20:05:24.442840 TCP connection from 10.78.11.124:50794 to 10.79.0.3:80 blocked by Weave NPC.
WARN: 2017/05/31 20:05:24.442898 TCP connection from 10.78.11.124:50794 to 10.79.96.1:80 blocked by Weave NPC.
WARN: 2017/05/31 20:05:27.458832 TCP connection from 10.78.11.124:50794 to 10.79.128.1:80 blocked by Weave NPC.
WARN: 2017/05/31 20:05:31.490880 TCP connection from 10.78.11.124:50794 to 10.79.96.1:80 blocked by Weave NPC.
```

Then I add the following network policy:

```yml
apiVersion: extensions/v1beta1
kind: NetworkPolicy
metadata:
  name: ingress-network-policy
  namespace: kube-system
spec:
  podSelector:
    matchLabels:
      name: traefik-ingress-lb
  ingress:
  - ports:
    - protocol: TCP
      port: 80
    - protocol: TCP
      port: 8080
```

... which basically lets everything through to 80 and 8080. npc catches the change, but nothing gets through:

```
INFO: 2017/05/31 20:19:43.604045 EVENT AddNetworkPolicy {"metadata":{"name":"ingress-network-policy","namespace":"kube-system","selfLink":"/apis/extensions/v1beta1/namespaces/kube-system/networkpolicies/ingress-network-policy","uid":"74fba4b1-463e-11e7-9600-028ec42709ca","resourceVersion":"11367","generation":1,"creationTimestamp":"2017-05-31T20:19:32Z","annotations":{"kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"extensions/v1beta1\",\"kind\":\"NetworkPolicy\",\"metadata\":{\"annotations\":{},\"name\":\"ingress-network-policy\",\"namespace\":\"kube-system\"},\"spec\":{\"ingress\":[{\"ports\":[{\"port\":80,\"protocol\":\"TCP\"},{\"port\":8080,\"protocol\":\"TCP\"}]}],\"podSelector\":{\"matchLabels\":{\"name\":\"traefik-ingress-lb\"}}}}\n"}},"spec":{"podSelector":{"matchLabels":{"name":"traefik-ingress-lb"}},"ingress":[{"ports":[{"protocol":"TCP","port":80},{"protocol":"TCP","port":8080}]}]}}
INFO: 2017/05/31 20:19:43.604138 creating ipset: &npc.selectorSpec{key:"name=traefik-ingress-lb", selector:labels.internalSelector{labels.Requirement{key:"name", operator:"=", strValues:[]string{"traefik-ingress-lb"}}}, ipsetType:"hash:ip", ipsetName:"weave-W[F1)yIM0$Dy9nhG~K%Jn*XR4"}
INFO: 2017/05/31 20:19:43.605463 adding rule: [-p TCP -m set --match-set weave-W[F1)yIM0$Dy9nhG~K%Jn*XR4 dst --dport 80 -j ACCEPT]
INFO: 2017/05/31 20:19:43.607732 adding rule: [-p TCP -m set --match-set weave-W[F1)yIM0$Dy9nhG~K%Jn*XR4 dst --dport 8080 -j ACCEPT]
DEBU: 2017/05/31 20:19:45.664912 UpdatePod ignored for pod default/nginx-2371676037-8827c on node 10.78.22.156
DEBU: 2017/05/31 20:19:46.693080 UpdatePod ignored for pod kube-system/kube-dns-321336704-2bb85 on node 10.78.22.156
WARN: 2017/05/31 20:19:47.060889 TCP connection from 10.78.12.11:32869 to 10.79.0.3:80 blocked by Weave NPC.
DEBU: 2017/05/31 20:19:52.377437 UpdatePod ignored for pod kube-system/kube-dns-321336704-2bb85 on node 10.78.22.156
WARN: 2017/05/31 20:20:07.411835 TCP connection from 10.78.11.124:50802 to 10.79.0.3:80 blocked by Weave NPC.
WARN: 2017/05/31 20:20:07.411869 TCP connection from 10.78.11.124:50802 to 10.79.96.1:80 blocked by Weave NPC.
```

FYI, before my latest upgrade to 1.9.6, traffic from outside got through. I always could disable npc, but I prefer to get it working correctly.

Is there an error in my networkpolicy? Or bug in npc?

To be honest, I always had thought that NetworkPolicy was primarily for pod-to-pod access, not external addresses - I don't see in the networkpolicypeer spec where you can select anything other than internal cluster elements (namespace or labels) - but clearly I was wrong, since weave-npc is blocking an address not in the cluster.

**Comment by deitch** [comment]
Do not know if this is related to https://github.com/weaveworks/weave/issues/2995 (yes, we override hostname) and https://github.com/weaveworks/weave/issues/2994

**Comment by bboreham** [comment]
In the log messages like `ignored for pod default/nginx-2371676037-8827c on node 10.78.22.156`, it's supposed to ignore every pod on other nodes, but not ignore pods on the local node.

It should print out as the very first line what it thinks the local node is called.  So if you catch it ignoring local nodes then this is the same issue as #2995 

#2994 is about over-reporting when things are not actually blocked, which is not the same symptom.

**Comment by deitch** [comment]
So what do I do with this line:

```
WARN: 2017/05/31 20:20:07.411835 TCP connection from 10.78.11.124:50802 to 10.79.0.3:80 blocked by Weave NPC.
```

It is blocking connections coming from off-cluster entirely.

**Comment by bboreham** [comment]
This could be caused by #2995

**Comment by bboreham** [comment]
Note that we changed the yaml in 1.9.7 to address this; are you using that version?

**Comment by deitch** [comment]
> Note that we changed the yaml in 1.9.7 to address this; are you using that version?


We used the yaml generator, so whatever came out of it. We submitted the issue 5 days ago, 31 May, so was that prior to 1.9.7?

Looking at the logs, it appears that we had 1.9.6, so it must have been before. Should we remove the daemonset and service and reapply to see if it helps?

**Comment by deitch** [comment]
@bboreham 

1. I upgraded to 1.9.7, still have the problem.
2. There is a bug in the generator. It puts the DaemonSet in default namespace rather than kibe-system. I will open a new issue for this one?

**Comment by deitch** [comment]
Add to that 3.

I used the following URL for the generator: curl "https://cloud.weave.works/k8s/v1.6/net.yaml?env.IPALLOC_RANGE=10.79.0.0/16&env.EXPECT_NPC=0"

I figured, might as well try disabling npc entirely. npc still starts up and still blocks everything from outside.

**Comment by deitch** [comment]
Issues raised for generator [here](https://github.com/weaveworks/weave/issues/3006) and [here](https://github.com/weaveworks/weave/issues/3007)

So we can leave this focused on the npc blocking issue.

**Comment by deitch** [comment]
Interesting, as a quick check. Docker logs for weave-npc:

```
core@ip-10-78-20-30 ~ $ docker logs -f 99567e69b2e4
INFO: 2017/06/05 11:23:26.676382 Starting Weaveworks NPC 1.9.7; node name "ip-10-78-20-30.ec2.internal"
INFO: 2017/06/05 11:23:26.677193 Serving /metrics on :6781
Mon Jun  5 11:23:26 2017 <5> ulogd.c:843 building new pluginstance stack: 'log1:NFLOG,base1:BASE,pcap1:PCAP'
```

note the node name is `"ip-10-78-20-30.ec2.internal"`.

Now kubelet process:

```
/opt/local/bin/kubelet --allow-privileged=true --cloud-provider= --hostname-override 10.78.20.30 --network-plugin=cni --cni-conf-dir=/etc/cni/net.d --cni-bin-dir=/opt/cni/bin --cluster-dns=10.80.0.5 --cluster-domain=cluster.local --container-runtime=docker --docker=unix:///var/run/docker.sock --kubeconfig=/var/lib/kubelet/kubeconfig --register-node=true --require-kubeconfig=true --serialize-image-pulls=false --tls-cert-file=/var/lib/kubernetes/kubernetes-worker.pem --tls-private-key-file=/var/lib/kubernetes/kubernetes-worker-key.pem --enable-cri=false --v=2
```

Override the hostname, weave-npc doesn't seem to know it.

**Comment by marccarre** [comment]
> We used the yaml generator, so whatever came out of it

I would **not** recommend using the Launch Generator with Weave Net 1+, but only with `master` or Weave Net 2.0, once released, which is why we haven't advertised the Launch Generator just yet. Indeed some of the defaults it sets could potentially break previous versions, and we do not plan to make it backward compatible: YAMLs from the GitHub releases page are there for this.

For example, since persistence will change in 2.0 you'd have:
- Launch Generator:
```
            - name: weavedb
              hostPath:
                path: /var/lib/weave
```
- 1.9.7's static YAMLs from the release page:
```
        - name: weavedb
          emptyDir: {}
```

**Comment by deitch** [comment]
> I would not recommend using the Launch Generator with Weave Net 1+

Bryan said in a different issue that I could use it (but of course I cannot find it now...)

Doesn't the generator determine the version? If my only entrypoint to launching weave _is_ the generator, then _it_ determines the version of container, does it not? Unless it is just taking `latest`.

Either way, I am pretty sure it failed even when I took the static git.io one and manually added the `IPALLOC_RANGE`. I will double-check now.

**Comment by deitch** [comment]
I confirm that it still happens with the static git.io script modified to add `IPALLOC_RANGE` . In that case, I am running 1.9.7.

**Comment by bboreham** [comment]
Believe this should be fixed by #3014

**Comment by deitch** [comment]
Thanks @bboreham . If it recurs, I will open a new issue.

**Comment by bboreham** [comment]
Note the fix is in 1.9.8 just released

**Comment by deitch** [comment]
And now I have the reverse issue: it appears *never* to block anything. Will open a new issue.

**Comment by deitch** [comment]
Is it possible this is still an issue on 2.2.0 (which is what I am running)? I have a kafka pod listening on 9092 and via a service on `nodePort 30094`. If I try to get to `host:30094` from a pod running on the same host, it works fine; from a pod running on another host, I get 

```
WARN: 2019/05/09 19:14:28.236915 TCP connection from 10.50.22.140:54624 to 10.51.24.5:9092 blocked by Weave NPC.
```

FWIW:

* `10.50.22.140` is the IP of host A on which the source pod is running
* `10.51.24.5` is the pod IP of the target pod on the host B
* `9092` is the pod port
* `30094` is the nodePort of the service

The following works:

* source pod to target pod IP `9092`
* target pod to target pod host IP `30094`

But source pod to target pod host IP `30094` is blocked as above.

Definitely running 2.2.0, yaml as generated by the generator (plus modified to add the `IPALLOC_RANGE` since we still have `--cluster-cidr` on kube-proxy).

Is there something else? Should I open a new issue?

**Comment by murali-reddy** [comment]
@deitch It is still an issue with even recent version, but I doubt that is something that can be fixed in weave-net.

So underlying problem is in this scenario, we are crossing overlay <-> underlay boundary i.e) pod accessing a remote node port, so weave-net  `MASQUERADE` the traffic. Hence you would see the  host IP on the destination node. For e.g. in your logs `10.50.22.140` is host IP.

But if you access the same node port on `weave` bridge IP of remote node it would work.

**Comment by deitch** [comment]
> So underlying problem is in this scenario, we are crossing overlay <-> underlay boundary i.e) pod accessing a remote node port, so weave-net MASQUERADE the traffic. Hence you would see the host IP on the destination node. For e.g. in your logs 10.50.22.140 is host IP

Yes, I figured as much. Optimally, pod A on host A would talk to pod B on host B using pod B's IP or, at most, the fronting service's IP, and not host B's IP and nodePort. 

The specific situation is kafka, where the protocol is not at all like a "normal" Web server. When you connect to Kafka on port 9092 (or any other non-standard port), you don't send messages. In stead it is a "bootstrap" type protocol, where Kafka will tell you, "thanks for connecting, here are the IPs and ports where you can reach brokers." If the broker itself needs to be reached both from inside and outside the k8s cluster, you have an issue: advertising the external IP (via service's nodePort) means traffic is blocked per this issue; advertising the internal IP means consumers/producers outside the cluster cannot reach Kafka. Even if you put a service with loadbalancer or nodeport or ingress in front of the kafka broker, it will advertise the internal port, and there is no way to produce or consume.

> I doubt that is something that can be fixed in weave-net.

I am not sure that is necessarily true. I can think of some ways to do it - e.g. having weave recognize that the target host and IP actually is a k8s worker and nodePort, so send it directly to the target pod - but it is non-trivial. On the other hand, most of k8s networking is built precisely like that: recognizing that the target service IP and port actually map to pod X on host Y and sending it direct. iptables fun.

> But if you access the same node port on weave bridge IP of remote node it would work.

I don't understand?

**Comment by deitch** [comment]
Also @murali-reddy , the earlier issue here and its resolution implies that it was fixed. I remember opening the original issue, and remember the discussion with @bboreham about it both here and on Slack, and am pretty sure I remember him explaining how it can and will be fixed, but relying on memory, so....

**Comment by murali-reddy** [comment]
>> But if you access the same node port on weave bridge IP of remote node it would work.
> 
> I don't understand?

If you do `ip addr list weave` you should see IP assigned to `weave` bridge in host namespace which is from pod subnet. You can access the nodeport on that IP as well.

**Comment by deitch** [comment]
Oh, that _is_ an interesting idea. It leads to 2 questons:

1. How would I make that known (in a programmatic fashion) to my kafka broker pods? Normally I just use the downward API.
2. Would that be reachable from outside the cluster? I know the host IP is, and is routed in a normal fashion by the cloud provider. What about this one?

**Comment by murali-reddy** [comment]
I just used it to describe crossing of overlay-underlay boundary is the problem here. IP assigned to `weave` bridge IP is not reachable from out side the cluster.

**Comment by deitch** [comment]
So it doesn't really provide an answer.

OK, let's turn it around. If you want to run kafka brokers on Kubernetes, accessible _only_ on 9092, and both from within and without the cluster, how would you do it? NetworkPolicy is the right way, this issue addresses the problem, it says it was fixed, but it isn't. Is that about right?

**Comment by murali-reddy** [comment]
It might be confusing to discuss on old issue. Perhaps we should start a new issue.

Network policy that you shared originally in the [comment](https://github.com/weaveworks/weave/issues/2997#issue-232689907) which is to allow all the traffic on specific ports works fine as expected. I don't know if it was not working on 2.2. But on latest releases it works fine.

**Comment by deitch** [comment]
Good point. I will open a new issue.

**Comment by deitch** [comment]
New issues opened, see #3642
